### PR TITLE
chore: introduce code coverage

### LINKED ISSUE #1375: Introduce code coverage
### Is your enhancement related to a problem? Please describe

To have a way how to notify the contributor about code coverage changes. 

### Describe the solution you'd like

Provide a contributor a way to visually check see changes in code coverage when adding new code. To be run individually and locally for now.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by odockal**
New script to be run using yarn, test:coverage.
It uses c8 provider, configuration is done via vitest config file. Not part of the CI. Aimed at important typescript files. Omits svelte files. Closes #1375

Signed-off-by: Ondrej Dockal <odockal@redhat.com>

### What does this PR do?
Introduces an option to run code coverage for the developer to visually better understand missing unit tests for new logic.

### Screenshot/screencast of this PR
```
...
  registries.ts                            |       0 |        0 |       0 |       0 | 1-73                                                                       
  run-image-store.ts                       |       0 |        0 |       0 |       0 | 1-26                                                                       
  search-util.ts                           |   89.13 |    82.35 |     100 |   89.13 | 29-30,40,44-45                                                             
  statusbar.ts                             |       0 |        0 |       0 |       0 | 1-36                                                                       
  volumes.ts                               |       0 |        0 |       0 |       0 | 1-81                                                                       
-------------------------------------------|---------|----------|---------|---------|----------------------------------------------------------------------------

=============================== Coverage summary ===============================
Statements   : 7.71% ( 1286/16674 )
Branches     : 46.23% ( 86/186 )
Functions    : 16.49% ( 47/285 )
Lines        : 7.71% ( 1286/16674 )
================================================================================
```

### What issues does this PR fix or reference?

Fixes #1375 

### How to test this PR?

```
yarn test:coverage
```
coverage report are placed undet `./test-resources/coverage`

### CODE REVIEW

File: vitest.config.js [hunk 1]
```diff
@@ -33,18 +34,37 @@ const config = {
 
     exclude: [
       '**/builtin/**',
-      '**/node_modules/**',
-      '**/dist/**',
       '**/cypress/**',
-      '**/.{idea,git,cache,output,temp}/**',
+      '**/.{output,temp,cdix}/**',
```
Comment (benoitf) [inline vitest.config.js]: why removing the upper excludes ?

File: extensions/crc/scripts/build.js [hunk 2]
```diff
@@ -19,13 +19,13 @@
 
 const zipper = require('zip-local');
 const path = require('path');
-const package = require('../package.json');
+const packageJson = require('../package.json');
```
Comment (benoitf) [inline extensions/crc/scripts/build.js]: I don't see why these changes are related to code coverage ?

File: vitest.config.js (hunk 1 above)
Comment (odockal) [reply vitest.config.js]: they are included in configDefaults.exclude var, but I was playing around, it is better to have them extracted, will revert.

Comment (benoitf) [reply vitest.config.js]: ok thanks

File: extensions/crc/scripts/build.js (hunk 2 above)
Comment (benoitf) [reply extensions/crc/scripts/build.js]: it can be a separate PR if you prefer to rename some constants but it should not be related to code coverage

Comment (odockal) [reply extensions/crc/scripts/build.js]: At the beginning, some of the coverage engines complained about keyword `package` being used, now it is gone with c8 coverage engine, so reverting that back.

**Comment by odockal** [comment]
@benoitf I just have a fail in test coverage report, it does not prevent its creation, it just omits feedback folder which has this error:
```
 FAIL  packages/renderer/src/lib/feedback/SendFeedback.spec.ts [ packages/renderer/src/lib/feedback/SendFeedback.spec.ts ]
ReferenceError: expect is not defined
 ❯ Object.<anonymous> node_modules/@testing-library/jest-dom/dist/extend-expect.js:9:1
 ❯ Object.<anonymous> node_modules/@testing-library/jest-dom/dist/index.js:3:1
```
seems that this is problem when coverage is run globally. Was trying to workaround it, or find a fix, but without a success.

**Comment by benoitf** [comment]
@odockal it means the wrong vitest config file is used (if you look at some tests we specify the custom vitest config file)

**Comment by odockal** [comment]
@benoitf So the actual state is to divide coverage to be run per module/folder (extension, packages/main,preload,preload-docker-extension,renderer). It can be run as part of `test:extensions` command/workflow, or I have added an option to run it all `test:coverage` except e2e which I believe is not important from unit tests coverage point of view. I would also propose next steps to this one, to came up with a merge step, since this only introduce multiple folders within coverage folder, then refactor config extraction as sonar complains about multiple same config used and finally, if there will be interest, use the reports for sonar reporting. wdyt?

**Comment by benoitf** [comment]
Hello @odockal your proposal is looking fine 👍

**Comment by odockal** [comment]
@benoitf final question. how is it with yarn.lock? Should I add it into PR as well? PR seems to complain about it.

**Comment by benoitf** [comment]
@odockal yes if you modify package.json you need to modify yarn.lock file (commit it as well)

**Comment by odockal** [comment]
> @odockal yes if you modify package.json you need to modify yarn.lock file (commit it as well)

I hope everything is at place now.

**Review by benoitf — APPROVED** [review]
thanks for the code coverage report being produced automatically

image

(and also on the IDE when you have the right extension/plug-in)
